### PR TITLE
support for providing credentials in environment variables

### PR DESCRIPTION
**Pull request by mpnowacki-reef**
depends on reef-technologies/b2-sdk-python#113

**README/help changes**
New features are described in command help texts, but not in the main README, because it didn't seem like such a good idea to go into this much details in it. I can add them though.

**Cache classes**
The already existsing classes InMemoryAccountInfo and InMemoryCache seem to do their job fine. The only problem I noticed is the method: get_bucket_name_or_none_from_allowed - it seems to work differently for
InMemoryAccountInfo and SqliteAccountInfo - but here's the trick, for SqliteAccountInfo it won't work - it's a piece of not working dead code. Should this be removed?

**parent parser**
I added the possibility to provide --dev/--staging/--environment=... to each CLI Command that can authorize implicitly, but I did it "by hand" instead of using `parents` in `get_parser` - this approach seemed more readable.

**SDK version**
master of this repo cannot be used with the master of SDK repo (at least according to requirements.txt) - Is there a schedule saying when the CLI will start using SDK 1.3.0? That's an issue here because this PR depends on a PR in SDK.

### CODE REVIEW

File: b2/console_tool.py [hunk 1]
```diff
@@ -198,7 +204,17 @@ def run(self, args):
 
     @classmethod
     def _setup_parser(cls, parser):
-        pass
+        if cls.REQUIRES_AUTH:
+            cls._environment_parser_args(parser)
+
+    @classmethod
+    def _environment_parser_args(cls, parser):
```
Comment (mlech-reef) [inline b2/console_tool.py:211]: We have a place for common arguments (search for e.g. `--verbose`). Why did you add something else to meet the same goal?

File: b2/console_tool.py [hunk 2]
```diff
@@ -297,21 +320,26 @@ class AuthorizeAccount(Command):
...
     FORBID_LOGGING_ARGUMENTS = True
+    REQUIRES_AUTH = False
 
     @classmethod
-    def _setup_parser(cls, parser):
-        realm_group = parser.add_mutually_exclusive_group()
-        realm_group.add_argument('--dev', action='store_true', help=argparse.SUPPRESS)
-        realm_group.add_argument('--staging', action='store_true', help=argparse.SUPPRESS)
-        realm_group.add_argument('--environment', help=argparse.SUPPRESS)
+    def name_and_alias(cls):
```
Comment (mlech-reef) [inline b2/console_tool.py:326]: I don't get it. Why?

**Comment by mlech-reef** [comment]
> README/help changes
For me, documentation is enough.

> Cache classes
Can you fix that?

> parent parser
See my comment

> SDK version
When it's merged to master it will work. Until then you can test it manually.

File: b2/console_tool.py (hunk 2 above)
Comment (mpnowacki-reef) [reply b2/console_tool.py:326]: I used this command's name in command docstrings and errors printed to stderr, so I wanted to have it parameterized, to avoid wrong help texts/errors should the name of the command be changed.

Comment (mlech-reef) [reply b2/console_tool.py:326]: Actually, maybe instead of adding it like that, we should have a third env var for setting up custom realm?

File: b2/console_tool.py (hunk 1 above)
Comment (mlech-reef) [reply b2/console_tool.py:211]: Actually, maybe instead of adding it like that, we should have a third env var for setting up custom realm?

File: b2/console_tool.py (hunk 2 above)
Comment (mlech-reef) [reply b2/console_tool.py:326]: The command name should not be parametrized itself. It should be fixed :) It worked before based on the class name and we don't need special treatment for this one imho

File: b2/console_tool.py (hunk 1 above)
Comment (mpnowacki-reef) [reply b2/console_tool.py:211]: that would make a lot of things easier, and it's only used by BB's internal testing, right? Question is: 3 env vars (similar to 3 CLI options) or one with `dev/staging/url` value? I think the most obvious way to go would be to provide a single var, checked before args - but that's against the usual flow of `args->envvars`

File: b2/console_tool.py (hunk 2 above)
Comment (mpnowacki-reef) [reply b2/console_tool.py:326]: ok, so wherever the string `authorize-account` is used, it'll just be used as a literal.

File: b2/console_tool.py (hunk 1 above)
Comment (mlech-reef) [reply b2/console_tool.py:211]: Wy before args? Can we just change `_get_realm` like:

```python
    @classmethod
    def _get_realm(cls, args):
        if args.dev:
            return 'dev'
        if args.staging:
            return 'staging'
        if args.environment:
            return args.environment

        return os.environ.get(B2_ENVIRONMENT_ENV_VAR, 'production')
```
?

Comment (mpnowacki-reef) [reply b2/console_tool.py:211]: that also works. Now in order to use it inside of other commands, maybe change `args.dev` to `getattr(args, 'dev', None)` etc. ?

File: b2/console_tool.py (hunk 2 above)
Comment (mlech-reef) [reply b2/console_tool.py:326]: Unless we can extend `DOC_STRING_DATA` to somehow support every command :D I would not special threat just single command

File: b2/console_tool.py (hunk 1 above)
Comment (mlech-reef) [reply b2/console_tool.py:211]: Ok, but do we need to make it public now as there will be no `--dev`, `--staging` for other commands? Just envs. Or you want to keep them for some reason? But `appKey` and `appKeyId` can not be passed via command-line except for `authorize-account`, so why the realm should be different?

Comment (mpnowacki-reef) [reply b2/console_tool.py:211]: We do not need to make it public, but running any command, e.g. `ls my-bucket` may lead to running `authorize-account` implicitly, and that calls `_get_realms`. Guess it's best to separate `AuthorizeAccount.run` into parts:
* one processing arguments, calling `_get_realm`, `getpass` etc.
* the other actually doing all the "running", accepting `applicationKeyId`, `applicationKey` and `realm` as string arguments.

The second one would then be public and called by `authorize_from_env` - that way no `args` monkey patching needs to be done.

Comment (mlech-reef) [reply b2/console_tool.py:211]: That would be really nice

File: b2/console_tool.py [hunk 3]
```diff
@@ -631,6 +656,7 @@ class DeleteKey(Command):
     @classmethod
     def _setup_parser(cls, parser):
         parser.add_argument('applicationKeyId')
+        super()._setup_parser(parser)
```
Comment (mlech-reef) [inline b2/console_tool.py:659]: Remove

File: b2/console_tool.py [hunk 4]
```diff
@@ -655,6 +681,7 @@ def _setup_parser(cls, parser):
         parser.add_argument('--noProgress', action='store_true')
         parser.add_argument('fileId')
         parser.add_argument('localFileName')
+        super()._setup_parser(parser)
```
Comment (mlech-reef) [inline b2/console_tool.py:684]: Remove

File: b2/console_tool.py [hunk 5]
```diff
@@ -244,6 +249,11 @@ class B2(Command):
     """
     This program provides command-line access to the B2 service.
 
+    There are two flows of authorization:
```
Comment (mlech-reef) [inline b2/console_tool.py:252]: Recently we updated the doc strings as they are now used for generating the documentation (for sphinx and for `--help` with `rst2ansi`). So you have to update few things here:

* add a blank line after this one
* add double backticks around `{NAME} authorize-account`
* add double backticks around `{B2_APPLICATION_KEY_ID_ENV_VAR}`
* add double backticks around `{B2_APPLICATION_KEY_ENV_VAR}`

File: b2/console_tool.py [hunk 6]
```diff
@@ -1618,6 +1654,21 @@ def run_command(self, argv):
...
             raise
 
+    def authorize_from_env(self, command_class):
+        if not command_class.REQUIRES_AUTH:
+            return 0
+        if B2_APPLICATION_KEY_ID_ENV_VAR not in os.environ:
+            return 0
+
+        key_id = os.environ[B2_APPLICATION_KEY_ID_ENV_VAR]
+        key = os.environ.get(B2_APPLICATION_KEY_ENV_VAR, '')
```
Comment (mlech-reef) [inline b2/console_tool.py:1664]: Why the default is an empty string? It will always fail in that case. In my opinion, if only one authorization env is provided, we should either:
* pass-through
* raise MissingAccountData

Not sure which one is better here - probably the second one

Comment (mpnowacki) [reply b2/console_tool.py:1664]: Or maybe printing to stderr `"B2_APPLICATION_KEY" environment variable not provided` and returning exit status `1` ? I don't think there's a need for an exception here (especially, because `MissingAccountData` is kinda meant for columns).

Comment (mpnowacki-reef) [reply b2/console_tool.py:1664]: Or maybe printing to stderr `"B2_APPLICATION_KEY" environment variable not provided` and returning exit status `1` ? I don't think there's a need for an exception here (especially, because `MissingAccountData` is kinda meant for columns).

Comment (mlech-reef) [reply b2/console_tool.py:1664]: That would work
